### PR TITLE
Support transforms along arbitrary axes with jax.numpy.fft

### LINKED ISSUE #1878: fftn on 2D array with axes=[0] applies FFT to wrong axis
I'm not sure how this snuck through our unit tests!

```python
import jax.numpy as jnp
import numpy as np

signal = np.ones((2, 2))
print('numpy result:\n', np.fft.fftn(signal, axes=[0]).real)
print('jax.numpy result:\n', jnp.fft.fftn(signal, axes=[0]).real)
```
Prints:
```
numpy result:
 [[2. 2.]
 [0. 0.]]
jax.numpy result:
 [[2. 0.]
 [2. 0.]]
```

### PR DESCRIPTION
**Pull request by shoyer**
Fixes GH-1878

The logic that attempted to check for transformations along non-innermost axes
was broken.

Rather than fixing it, this PR adds support for these transformations by
transposing and untransposing arrays. This adds some overhead over the LAX
implementation, but it suspect it is minimal in most cases and it should be
worthwhile for the sake of completeness.

### CODE REVIEW

File: tests/fft_test.py [hunk 1]
```diff
@@ -90,20 +94,16 @@ def testFftnErrors(self, inverse):
     func = np.fft.ifftn if inverse else np.fft.fftn
     self.assertRaisesRegex(
         ValueError,
-        "jax.np.fft.{} only supports 1D, 2D, and 3D FFTs over the innermost axes. "
-        "Got axes None with input rank 4.".format(name),
+        "jax.np.fft.{} only supports 1D, 2D, and 3D FFTs".format(name),
```
Comment (skye) [inline tests/fft_test.py]: nit: I like having the full error message in tests (within reason). It makes it easy to review that the error message is clear and doesn't have typos (for both readers and authors changing the error message). Up to you though.

File: jax/numpy/fft.py [hunk 2]
```diff
@@ -60,15 +60,24 @@ def _fft_core(func_name, fft_type, a, s, axes, norm):
...
 
-  if any(axis in range(a.ndim - 3) for axis in axes):
+  if len(axes) > 3:
     raise ValueError(
-        "%s only supports 1D, 2D, and 3D FFTs over the innermost axes."
-        " Got axes %s with input rank %s." % (full_name, orig_axes, a.ndim))
+        "%s only supports 1D, 2D, and 3D FFTs. "
+        "Got axes %s with input rank %s." % (full_name, orig_axes, a.ndim))
+
+  if orig_axes is not None:
```
Comment (skye) [inline jax/numpy/fft.py]: Comment that XLA only supports up to 3D FFTs over the innermost axes, so we rearrange if necessary (otherwise it's pretty mysterious why we would do this).

**Review by skye — APPROVED** [review]
Nice! My apologies for the buggy test code...
